design/cli-install-changes.md

```diff
@@ -0,0 +1,426 @@
+
+#### GitOps Compatibility
+
+To maintain compatibility with gitops practices, each of the new commands will generate `yaml` output that can be stored in source control.
+
+For content examples, please refer to the files here:
+
+https://github.com/carlisia/velero/tree/c-cli-design/design/CLI/PoC
+
+Note: actual `yaml` file names are defined by the user.
+
+`velero config server` - base/deployment.yaml
+
+`velero config restic` - overlays/plugins/restic.yaml
+
+`velero backup-location create` - base/backupstoragelocations.yaml
+
+`velero snapshot-location create` - base/volumasnapshotlocations.yaml
+
+`velero plugin add velero/velero-plugin-for-aws:v1.0.1` - overlays/plugins/aws-plugin.yaml
+
+`velero plugin add velero/velero-plugin-for-microsoft-azure:v1.0.1` - overlay/plugins/azure-plugin.yaml
+
+These files can be deployed using the included kustomize setup by running `k apply -k design/CLI/PoC/overlays/plugins/`.
+
+Note: All CRDs, including the `ResticRepository`, may continue to be deployed at startup as it is now, or together with their respective instantiation.
+
+
+#### Changes to startup behavior
+
+To recap, this proposal redesigns the Velero CLI to make `velero install` obsolete, and instead breaks down the installation and configuration into separate commands. These are the major highlights:
+
+- Plugins will only be installed separately via `velero plugin add`
+- BSL/VSL will be continue to be configured separately, and now each will have an associated secret
+
+Since each BSL/VSL will have its own association with a secret, the user will no longer need to upload a new secret whenever changing to, or adding, a BSL/VSL for a provider that is different from the one in use. This will be done at setup time. This will make it easier to support any number of BSL/VSL combinations, with different providers each.
+
+The user will start up the Velero server on a cluster by using the command `velero config server`. This will create the Velero deployment resource with default values or values overwritten with flags, create the Velero CRDs, and anything else that is not specific to plugins or BSL/VSL.
+
+The Velero server will start up, verify that the deployment is running, that all CRDs were found, and log a message that it is waiting for a BSL to be configured. at this point, other operations, such as configuring restic, will be allowed. Velero should keep track of its status, ie, if it is ready to create backups or not. This could be a field `ServerStatus` added to `ServerStatusRequest`. Possible values could be [ready|waiting]. "ready" would mean there is at least 1 valid BSL, and "waiting" would be anything but that.
```

Comment:
We should probably try to use Kubernetes standards here and set up a liveness/readiness probe that signals Velero's readiness.

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes

If these fail, then the pod won't even be marked as `Ready` so tools like `kubectl` will report status, in addition to our own.

Reply:
https://github.com/vmware-tanzu/velero/pull/2382 is an example of an endpoint that could be consumed for liveness.

Readiness in kubernetes is used to route network traffic; the pod will be left alone if the readiness check fails, but no network traffic will be sent to it.

Liveness checks will restart the pod if they fail. This means we can lean on Kubernetes's abilities to manage the restarting/retrying for us, so long as we have an endpoint that accurately represents what we want.